crates/autopilot/src/domain/competition/winner_selection/combinatorial.rs

```diff
@@ -0,0 +1,354 @@
+        eth::{self, WrappedNativeToken},
+        fee,
+        settlement::{
+            math,
+            transaction::{self, ClearingPrices},
+        },
+    },
+    anyhow::{Context, Result},
+    ethcontract::U256,
+    itertools::Itertools,
+    std::collections::{HashMap, HashSet},
+};
+
+impl Arbitrator for Config {
+    fn filter_unfair_solutions(
+        &self,
+        mut participants: Vec<Participant<Unranked>>,
+        auction: &domain::Auction,
+    ) -> Vec<Participant<Unranked>> {
+        // Discard all solutions where we can't compute the aggregate scores
+        // accurately because the fairness guarantees heavily rely on them.
+        let scores_by_solution = scores_by_solution(&mut participants, auction);
+        participants.sort_unstable_by_key(|participant| {
+            std::cmp::Reverse(participant.solution().score().get().0)
+        });
+        let baseline_scores = compute_baseline_scores(&scores_by_solution);
+        participants.retain(|p| {
+            let aggregated_scores = scores_by_solution
+                .get(&SolutionKey {
+                    driver: p.driver().submission_address,
+                    solution_id: p.solution().id(),
+                })
+                .expect("every remaining participant has an entry");
+            // only keep solutions where each order execution is at least as good as
+            // the baseline solution
+            aggregated_scores.iter().all(|(pair, score)| {
+                baseline_scores
+                    .get(pair)
+                    .is_none_or(|baseline| score >= baseline)
+            })
+        });
+        participants
+    }
+
+    fn mark_winners(&self, participants: Vec<Participant<Unranked>>) -> Vec<Participant> {
+        let winners = self.pick_winners(participants.iter().map(|p| p.solution()));
+        participants
+            .into_iter()
+            .enumerate()
+            .map(|(index, participant)| participant.rank(winners.contains(&index)))
+            .collect()
+    }
+
+    fn compute_reference_scores(
+        &self,
+        participants: &[Participant],
+    ) -> HashMap<eth::Address, Score> {
+        let mut reference_scores = HashMap::default();
+
+        for participant in participants {
+            let solver = participant.driver().submission_address;
+            if reference_scores.len() >= self.max_winners {
+                // all winners have been processed
+                return reference_scores;
+            }
+            if reference_scores.contains_key(&solver) {
+                // we already computed this solver's reference score
+                continue;
+            }
+
+            let solutions_without_solver = participants
+                .iter()
+                .filter(|p| p.driver().submission_address != solver)
+                .map(|p| p.solution());
+
+            let winners = self.pick_winners(solutions_without_solver.clone());
```

Comment:
Without diving into details, it is not obvious that winners are just indices.
```suggestion
            let winner_idxs = self.pick_winners(solutions_without_solver.clone());
```